### PR TITLE
Isabel Suchanek & Jessica Owens -- Carets

### PR DESCRIPTION
**Pull request by isabeldepapel**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  Because it doesn't make sense to have to instantiate an instance of the class just to be able to call the score method to score the word. 'Score' isn't really something a particular Scoring object would do--it's something that really relates to the whole class   |
|   Describe an example of a test you wrote for a _nominal case_. |   In our dictionary spec, after we made sure that a Dictionary instance was being created, and it was set up as a hash, we wrote a test to make sure our binary search for the word would find a word that existed in the dictionary, and return true if it was a word (and false if it wasn't in the dictionary and thus not a word). We tested to see if our is_word? method returned true if passed the word "azotic" (which was in the dictionary). 	|
|   Describe an example of a test you wrote for an _edge case_. |   In the scoring spec, we have a series of tests that test improper input for self.score. Our method was set up so that it would raise an error if passed improper input (e.g. an empty string, not a string), so we used procs in our test to make sure an error was being raised when we passed self.score an empty string, nil, and non-string values.	|
|   What was your final test coverage percentage?	|   81.03%  But that's because we had a hard time writing tests for our game and board classes (especially the game class), which really brought our percentage down. The coverage for most of our code aside from game & board is pretty good (and the only reason player doesn't have 100% coverage, I think, is because of the private method which we can't test).	|
|   Is there a specific piece of code you'd like feedback on?	|   We ran out of time before we could write it, but we were trying to think of a way to place tiles on the board while being able to check if the user was playing a word or adding to a word that was already on the board. Here's our pseudocode for it--we're mostly wondering if this would be the right way to go:  We have the word and its length. We have the position of where its first letter would go (on the board) as a row and col index and its direction (across or down). Go through each index on the board and see if there is already a tile there AND if that tile corresponds to the correct letter for that position in the word. If no -- choose a different word OR pass. If yes -- remove the letters on the board from the required letters to have and then check player's tiles for remaining letters	|

### CODE REVIEW

**Comment by tildeee** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  x  |
|   Both Teammates contributed to the codebase	|  x  |
|   Regular Commits with meaningful commit messages  |  x  |
|  **Scoring class**	|   	|
|  `score` method	|  x  |
|   Uses appropriate data structure to store the letter score	|  x  |
|   Appropriately handles edge cases	|  x  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  x  |
|  `highest_scoring_word` method	|  x  |
|   Appropriately handles edge cases	|  x  |
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|  doesn't check for nothing passed in  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  x  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  x  |
|  Returns `true` or `false` from `won?` method |  x  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |    |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  x  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  x  |
| `tiles_remaining` method returns a count	|  x  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|    |

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        return
+      else
+        puts "What word do you want to play?"
+        word = gets.chomp
+
+        # check if a valid word
+        while !@dictionary.is_word?(word)
+          puts "Not a valid word. Do you want to play another word?"
+          answer = y_or_n
+          if answer == y
```
Comment (tildeee) [inline lib/game.rb:80]: typo on `y`!

**Comment by tildeee** [comment]
Ah! Amazing!
Great job with the dictionary and game.
I actually couldn't find a way to get your board to show up in my Terminal; let's meet up sometime so I can get that to work on my machine.

I think the code is nice and simple and minimal and effective :) Maybe clean up some comments before you turn this in so you can see how sparkly and clean your code is! ✨ 


Re: feedback table
You two didn't formally check for some edge cases, like input for "&*#@"s etc
In terms of the pseudo code you provided: sounds good to me!

File: lib/dictionary.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+      # check for empty string
+      return false if word.length < 1
+
+      letter1 = word[0]
+
+      # get all words beginning with first letter from dict
+      words_beginning_with_letter1 = @dictionary[letter1.to_sym]
+
+      # use binary search to search for word in the array
```
Comment (tildeee) [inline lib/dictionary.rb:44]: SWEEEEEET ✨
